File: codes/python/113.路径总和-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
	def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
		path = []
		res = []
		self.dfs(root, targetSum, path, res)
		return res
	
	def dfs(self, root, targetSum, path, res):
		if not root:
			return
		path.append(root.val)
		if not root.left and not root.right and targetSum == root.val:
			res.append(path[:])
		self.dfs(root.left, targetSum - root.val, path, res)
		self.dfs(root.right, targetSum - root.val, path, res)
		path.pop()

File: codes/java/113.路径总和-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
	public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
		List<List<Integer>> res = new ArrayList<>();
		List<Integer> path = new ArrayList<>();
		dfs(root, targetSum, res, path);
		return res;
	}

	private void dfs(TreeNode root, int targetSum, List<List<Integer>> res, List<Integer> path) {
		if (root == null) {
			return;
		}
		path.add(root.val);
		targetSum -= root.val;
		if (root.left == null && root.right == null && targetSum == 0) {
			res.add(new ArrayList<>(path));
		}
		dfs(root.left, targetSum, res, path);
		dfs(root.right, targetSum, res, path);
		path.remove(path.size() - 1);
	}
}